packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -94,19 +94,22 @@ function h(sel: string, data: VElementData, children: VNodes = EmptyArray): VEle
         });
     }
 
-    let elm;
-    const { key } = data;
+    const { key, ref } = data;
 
-    return {
+    const vnode: VElement = {
         type: VNodeType.Element,
         sel,
         data,
         children,
-        elm,
+        elm: undefined,
         key,
         hook: ElementHook,
         owner: vmBeingRendered,
     };
+
+    setRef(vmBeingRendered, ref, vnode);
```

Comment:
FYI. This approach will prevent us from hoisting any VNodes with a ref assigned to them.

Reply:
Good point. I need to look into how Vue does it to see if we can avoid that.

Reply:
I think I understand the problem... let me think about it.

Reply:
I did some digging and Vue handles `refs` during patching. So theoretically the VNode is your previous example can be hoisted.

https://github.com/vuejs/core/blob/5898629d723e82b68e9b17b91bf8b1a8390a3912/packages/runtime-core/src/renderer.ts#L465-L468

Also, Vue reuses the same ref object between render cycles. That said in dev mode, Vue returns a shallow read-only version of the underlying ref object:

https://github.com/vuejs/core/blob/1574edd490bd5cc0a213bc9f48ff41a1dc43ab22/packages/runtime-core/src/componentPublicInstance.ts#L235